### DISCUSSION TITLE
Virtual character position

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#200] [Q&A] [answered]
I noticed that sometimes when I get character position via `body.GetPosition()`, I get a vector with NaN values. It happens so rarely, that I can't figure out the exact cause of it.

The only way I set the position of a virtual character is occasionally (on respawn) via:
```js
body.SetPosition(vec);
```
and mostly via
```js
body.SetLinearVelocity(vec);
```

Thinking about it, do I need to use body interface for setting position of a character instead? e.g. 
```js
bodyInterface.SetPositionAndRotation(body.GetID(), vec, quat, Jolt.EActivation_Activate);
```

**Comment by jrouwe** [comment]
If `body` is of type `CharacterVirtual` then `SetPosition` is the correct call. `SetPositionAndRotation` is for rigid bodies, not for characters.

Other than that it's hard to say what your problem is. I would recommend building a debug build and running with that. With a little bit of luck you will trigger an assert with a bit of extra information.

**Reply by LeXXik** [reply]
Thanks! Using a debug build did help some, triggering some asserts.

The scenario: a virtual character with a convex hull shape moving around and collecting kinematic coins on contact.

There were 2 asserts:

https://github.com/jrouwe/JoltPhysics/blob/81c98592a50e4c8d74415175ec7091c537045377/Jolt/Physics/Body/MotionProperties.cpp#L60
This one was due to artist creating the coins as a kinematic body with all degrees of freedom axes disabled (dofs == 0). Assert indicated that it may cause division by zero down the line. I am hoping that was the actual cause, as that should explain NaN for position/linvel for the character. Going to keep an eye on it still, as the issue was hard to reproduce in the first place.

https://github.com/jrouwe/JoltPhysics/blob/81c98592a50e4c8d74415175ec7091c537045377/Jolt/Physics/Character/CharacterVirtual.cpp#L547
That one triggers every step for all characters. I think it is due to them having convex hulls and the correction not being available. This doesn't seem to produce NaN, though, so I assume it is a "safe" assert.

**Comment by jrouwe** [comment, marked as answer]
> This one was due to artist creating the coins as a kinematic body with all degrees of freedom axes disabled (dofs == 0). Assert indicated that it may cause division by zero down the line. I am hoping that was the actual cause, as that should explain NaN for position/linvel for the character. Going to keep an eye on it still, as the issue was hard to reproduce in the first place.

Yes, this is quite likely to cause a NaN in position down the line.

> That one triggers every step for all characters. I think it is due to them having convex hulls and the correction not being available. This doesn't seem to produce NaN, though, so I assume it is a "safe" assert.

Convex hulls should be covered by the `if (inShape->GetType() == EShapeType::Convex)` branch. What is the exact shape type you're using for your characters? (maybe it's a convex hull wrapped in some other shape?) This assert can lead to the character getting stuck.

**Reply by LeXXik** [reply]
Looked into it, so the character shapes were wrapped like so:

```js
let settings = new Jolt.ConvexHullShapeSettings();
settings = new Jolt.RotatedTranslatedShapeSettings(vec, quat, settings);
settings = new Jolt.ScaledShapeSettings(settings, vec);
```

Since the scale was a unit vector anyway, I removed the scale wrap, and the assert is gone as well.

**Reply by LeXXik** [reply]
Thinking about it, the scaled wrap should be before rotatedTranslated in general? Or it doesn't matter?

**Reply by jrouwe** [reply]
It doesn't matter. I simply didn't add support for scaled shapes in CharacterVirtual. It is easy to add though:

https://github.com/jrouwe/JoltPhysics/pull/1283

**Reply by jrouwe** [reply]
Release 0.26.0 contains the changes
